source/common/common/assert.h

```diff
@@ -134,6 +165,38 @@ void invokeDebugAssertionFailureRecordAction_ForAssertMacroUseOnly();
     abort();                                                                                       \
   } while (false)
 
+#if !defined(NDEBUG)
+#define ENVOY_BUG_ACTION abort()
+#else
+#define ENVOY_BUG_ACTION Envoy::Assert::invokeEnvoyBugFailureRecordActionForEnvoyBugMacroUseOnly()
+#endif
+
+// These macros are needed to stringify __LINE__ correctly.
+#define STRINGIFY(X) #X
+#define TOSTRING(X) STRINGIFY(X)
+
+// CONDITION_STR is needed to prevent macros in condition from being expected, which obfuscates
+// the logged failure, e.g., "EAGAIN" vs "11".
+// ENVOY_BUG logging and actions are invoked only on power-of-two instances per log line.
+#define _ENVOY_BUG_IMPL(CONDITION, CONDITION_STR, ACTION, DETAILS)                                 \
+  do {                                                                                             \
+    if (!(CONDITION) && Envoy::Assert::shouldLogAndInvokeEnvoyBugForEnvoyBugMacroUseOnly(          \
+                            __FILE__ ":" TOSTRING(__LINE__))) {                                    \
```

Comment:
We have to make sure that windows build uses appropriate compiler flags, as by default __FILE__ does not have full path, just the filename. @sunjayBhatia @wrowe

Reply:
Is this issue of the lack of uniqueness in the contents of __FILE__ an argument to re-consider use of static atomics for these counters?

Reply:
Wouldn't counters have the same issue, since they'd be initialized according to file/line

Reply:
the regular component/misc logging uses `__FILE__` and shows the full path on Windows:
```
[2020-06-16 18:13:24.358][14656][debug][misc] [source/common/network/io_socket_error_impl.cc:30] Unknown error code 123 details...
```

Reply:
Can't find specifically where we pass `/FC` to MSVC, but seems that it is happening